lib/h3.rb

```diff
@@ -124,4 +130,85 @@ def self.h3_unidirectional_edges_from_hexagon(origin)
     getH3UnidirectionalEdgesFromHexagon(origin, edges)
     edges.read_array_of_ulong_long(max_edges).reject { |i| i == 0 }
   end
+
+  def self.h3_to_geo_boundary(h3_index)
+    geo_boundary = Structs::GeoBoundary.new
+    h3ToGeoBoundary(h3_index, geo_boundary)
+    geo_boundary[:verts].take(geo_boundary[:num_verts] * 2).map do |d|
+      rads_to_degs(d)
+    end.each_slice(2).to_a
+  end
+
+  def self.hex_ranges(h3_set, k)
+    h3_set.uniq!
+    max_out_size = h3_set.size * max_kring_size(k)
+    out = FFI::MemoryPointer.new(H3_INDEX, max_out_size)
+    pentagonal_distortion = false
+    FFI::MemoryPointer.new(H3_INDEX, h3_set.size) do |h3_set_ptr|
+      h3_set_ptr.write_array_of_ulong_long(h3_set)
+      pentagonal_distortion = hexRanges(h3_set_ptr, h3_set.size, k, out)
+    end
+    raise(ArgumentError, "One of the specified hexagon ranges contains a pentagon") if pentagonal_distortion
+
+    h3_range_indexes = out.read_array_of_ulong_long(max_out_size)
+    out = h3_set.inject({}) { |acc, i| acc[i] = []; acc }
+    h3_set.each_with_index do |h3_index, i|
+      (k + 1).times { out[h3_index] << [] }
+      0.upto(h3_set.count * max_kring_size(k) / h3_set.count).map do |j|
+        ring_index = ((1 + Math.sqrt(1 + 8 * (j / 6.0).ceil)) / 2).floor - 1
+        out[h3_index][ring_index] << h3_range_indexes[(i * h3_set.count + j)]
+        out[h3_index][ring_index].compact!
+      end
+      out[h3_index] = out[h3_index].sort_by(&:count)
+    end
+    out
+  end
+
+  def self.hex_range_distances(h3_index, k)
+    max_out_size = max_kring_size(k)
+    out = FFI::MemoryPointer.new(H3_INDEX, max_out_size)
+    distances = FFI::MemoryPointer.new(:int, max_out_size)
+    pentagonal_distortion = hexRangeDistances(h3_index, k, out, distances)
+    raise(ArgumentError, "Specified hexagon range contains a pentagon") if pentagonal_distortion
+
+    hexagons = out.read_array_of_ulong_long(max_out_size)
+    distances = distances.read_array_of_int(max_out_size)
+
+    Hash[
+      distances.zip(hexagons).group_by { |distance, _hexagon| distance }.map do |k, v|
+        [k, v.map { |_distance, hexagon| hexagon }]
+      end
+    ]
+  end
+
+  def self.k_ring_distances(h3_index, k)
+    max_out_size = max_kring_size(k)
+    out = FFI::MemoryPointer.new(H3_INDEX, max_out_size)
+    distances = FFI::MemoryPointer.new(:int, max_out_size)
+    kRingDistances(h3_index, k, out, distances)
+
+    hexagons = out.read_array_of_ulong_long(max_out_size)
+    distances = distances.read_array_of_int(max_out_size)
+
+    Hash[
+      distances.zip(hexagons).group_by { |distance, _hexagon| distance }.map do |k, v|
+        [k, v.map { |_distance, hexagon| hexagon }]
+      end
+    ]
+  end
+
+  def self.max_uncompact_size(hexagons, resolution)
+    FFI::MemoryPointer.new(H3_INDEX, hexagons.size) do |hexagons_ptr|
+      hexagons_ptr.write_array_of_ulong_long(hexagons)
+      return maxUncompactSize(hexagons_ptr, hexagons.size, resolution)
```

Comment:
I see `hexRanges` had a different approach. This code has a `return` inside the block (quits the method), but `hexRanges` set an outer flag and waited until the block was over to `raise`.

Is there a reason for the difference? If not, maybe pick one style?

Reply:
Yeah, `hexRanges` was quite hard to write and I'm actually considering just calling `hex_range_distances` since it returns the same data but in a form that's trivial to group.

May be some performance considerations for larger inputs though.

All of the code I've written so far is quite rough and will need refactoring heavily. I'm aiming to get all the functions wrapped with minimal specs first.